app/server/lib/commonUrls.ts

```diff
@@ -0,0 +1,81 @@
+import { ICommonUrls } from "app/common/ICommonUrls";
+
+import * as t from "ts-interface-checker";
+
+import ICommonUrlsTI from 'app/common/ICommonUrls-ti';
+
+const { ICommonUrls: ICommonUrlsChecker } = t.createCheckers(ICommonUrlsTI);
+
+function withAdminDefinedUrls(defaultUrls: ICommonUrls): ICommonUrls {
+  const adminDefinedUrlsStr = process.env.GRIST_CUSTOM_COMMON_URLS;
+  if (!adminDefinedUrlsStr) {
+    return defaultUrls;
+  }
+
+  let adminDefinedUrls;
+  try {
+    adminDefinedUrls = JSON.parse(adminDefinedUrlsStr);
+  } catch(e) {
+    throw new Error("The JSON passed to GRIST_CUSTOM_COMMON_URLS is malformed");
+  }
+
+  const merged = {
+    ...defaultUrls,
+    ...(adminDefinedUrls)
+  };
+  ICommonUrlsChecker.strictCheck(merged);
```

Comment:
BTW, is it not a problem that `ICommonUrlsChecker` makes every property required? Does it mean that you can't override only some, but must override all or none? This might be intentional, if you want to make sure you don't use any default values, but doesn't it interfere with upgrades whenever a new property gets added?

Reply:
> Does it mean that you can't override only some, but must override all or none?

No. If you read the code above, the user-defined properties are merged with the default ones, and that's the result that is checked here.

Which means that it fails in either cases:
 - `defaultUrls` object does not fully implement the `ICommonUrls` interface, but the error should occur during the build
 - `adminDefinedUrls` add properties that is not defined in the `ICommonUrls` interface, preventing the server to start.

Reply:
Oh you are right, sorry. I guess the second point may still be an issue -- it would cause failures on rare occasions that a property is removed; also if you add a new property to your config for a new version, the older versions wouldn't start. `ICommonUrlsChecker.check(merged)` would accept extra properties.